File: 2021/383. Ransom Note.cpp
class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        vector<int>count(26, 0);
        for (char c : ransomNote) {
            count[c - 'a']++;
        }
        for (char c : magazine) {
            count[c - 'a']--;
        }
        for (int n : count) {
            if (n > 0) return false;
        }
        return true;
    }
};

class Solution {
public:
    bool canConstruct(string ransomNote, string magazine) {
        unordered_map<int, int> mCount;
        for (char c : magazine) {
            mCount[c]++;
        }
        for (char c : ransomNote) {
            mCount[c]--;
            if (mCount[c] < 0) {
                return false;
            }
        }
        return true;
    }
};

File: 2024/383. Ransom Note.py
class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        magazine_count = defaultdict(int)
        for c in magazine:
            magazine_count[c] += 1
        ransomNote_count = defaultdict(int)
        for c in ransomNote:
            ransomNote_count[c] += 1
        for c, count in ransomNote_count.items():
            if magazine_count[c] < count:
                return False
        return True
